File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    answer = 0
    dx = [1,-1,0,0]
    dy = [0,0,1,-1]
    q = deque()
    q.append((0,0,1))
    def bfs(x,y):
        result = 0
        while q:
            mx,my,result = q.popleft()
            if mx == len(maps)-1 and my == len(maps[0])-1:
                return result
            for i in range(4):
                nx = mx + dx[i]
                ny = my + dy[i]
                if 0<= nx < len(maps) and 0<= ny < len(maps[0]) and maps[nx][ny] == 1:
                    maps[nx][ny] = 0
                    q.append((nx,ny,result+1))
        return -1
    return bfs(0,0)

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = -1;
    var q = [];
    var dx = [1,-1,0,0];
    var dy = [0,0,1,-1];
    
    q.push(0,0,1);
    
    while(q.length > 0){
        var x = q.shift();
        var y = q.shift();
        var fee = q.shift();
        
        for(let i = 0;i<4;i++){
            var nx = x + dx[i];
            var ny = y + dy[i];
            if(nx === maps.length-1 && ny === maps[nx].length){
                return fee;
            }
            if(nx >= 0 && nx < maps.length && ny >= 0 && ny < maps[nx].length && maps[nx][ny] === 1){
                maps[nx][ny] = 0;
                q.push(nx,ny,fee + 1);
            }
        }
    }
    return answer;
}